libraries/AP_Radio/AP_Radio_cc2500.cpp

```diff
@@ -0,0 +1,874 @@
+    cc2500.WriteReg(CC2500_0A_CHANNR, channel);
+    cc2500.Strobe(CC2500_SRX);
+}
+
+void AP_Radio_cc2500::nextChannel(uint8_t skip)
+{
+    channr = (channr + skip) % listLength;
+    setChannel(bindHopData[channr]);
+}
+
+void AP_Radio_cc2500::parse_frSkyX(const uint8_t *packet)
+{
+    uint16_t c[8];
+
+    c[0] = (uint16_t)((packet[10] <<8)& 0xF00) | packet[9];
+    c[1] = (uint16_t)((packet[11]<<4)&0xFF0) | (packet[10]>>4);
+    c[2] = (uint16_t)((packet[13] <<8)& 0xF00) | packet[12];
+    c[3] = (uint16_t)((packet[14]<<4)&0xFF0) | (packet[13]>>4);
+    c[4] = (uint16_t)((packet[16] <<8)& 0xF00) | packet[15];
+    c[5] = (uint16_t)((packet[17]<<4)&0xFF0) | (packet[16]>>4);
+    c[6] = (uint16_t)((packet[19] <<8)& 0xF00) | packet[18];
+    c[7] = (uint16_t)((packet[20]<<4)&0xFF0) | (packet[19]>>4);
+
+    uint8_t j;
+    for (uint8_t i=0;i<8;i++) {
+        if(c[i] > 2047)  {
+            j = 8;
+            c[i] = c[i] - 2048;
+        } else {
+            j = 0;
+        }
+        if (c[i] == 0) {
+            continue;
+        }
+        uint16_t word_temp = (((c[i]-64)<<1)/3+860);
+        if ((word_temp > 800) && (word_temp < 2200)) {
+            uint8_t chan = i+j;
+            if (chan < CC2500_MAX_CHANNELS) {
+                pwm_channels[chan] = word_temp;
+                if (chan >= chan_count) {
+                    chan_count = chan+1;
+                }
+            }
+        }
+    }
+}
+
+uint16_t AP_Radio_cc2500::calc_crc(uint8_t *data, uint8_t len)
+{
+    uint16_t crc = 0;
+    for(uint8_t i=0; i < len; i++) {
+        crc = (crc<<8) ^ (CRCTable[((uint8_t)(crc>>8) ^ *data++) & 0xFF]);
+    }
+    return crc;
+}
+
+bool AP_Radio_cc2500::check_crc(uint8_t ccLen, uint8_t *packet)
+{
+    uint16_t lcrc = calc_crc(&packet[3],(ccLen-7));
+    return ((lcrc >>8)==packet[ccLen-4] && (lcrc&0x00FF)==packet[ccLen-3]);
+}
+
+/*
+  save bind info
+ */
+void AP_Radio_cc2500::save_bind_info(void)
+{
+    // access to storage for bind information
+    StorageAccess bind_storage(StorageManager::StorageBindInfo);
+    struct bind_info info;
+    
+    info.magic = bind_magic;
+    info.bindTxId[0] = bindTxId[0];
+    info.bindTxId[1] = bindTxId[1];
+    info.bindOffset = bindOffset;
+    info.listLength = listLength;
+    memcpy(info.bindHopData, bindHopData, sizeof(info.bindHopData));
+    bind_storage.write_block(0, &info, sizeof(info));
+}
+
+/*
+  load bind info
+ */
+bool AP_Radio_cc2500::load_bind_info(void)
+{
+    // access to storage for bind information
+    StorageAccess bind_storage(StorageManager::StorageBindInfo);
+    struct bind_info info;
+
+    if (!bind_storage.read_block(&info, 0, sizeof(info)) || info.magic != bind_magic) {
+        return false;
+    }
+    
+    bindTxId[0] = info.bindTxId[0];
+    bindTxId[1] = info.bindTxId[1];
+    bindOffset = info.bindOffset;
+    listLength = info.listLength;
+    memcpy(bindHopData, info.bindHopData, sizeof(bindHopData));
```

Comment:
After blindly memcpy()-ing a chunk of data from ( possiby unreliable) storage into this struct, is it worth defensively checking to make sure that the data is 'good' before we trust it?